### DISCUSSION TITLE
Sensitivity Analysis with the simulation of multiple parameters.

### DISCUSSION THREAD

**Discussion by pvhau169** [casadi/casadi#3329] [Q&A]
Hi I'm having a problem with doing sensitivity analysis on the simulation of multiple parameters. I still don't understand how to use the parameter "fwd_p" of Integrator.factory(). I adapted the sensitivity analysis from https://github.com/casadi/casadi/blob/main/docs/examples/python/sensitivity_analysis.py which is using simulation of 1 parameter. 

```python
import casadi as cs
import numpy as np
import copy
import matplotlib.pyplot as plt

def simplerSim(x, t, k):
    return cs.vertcat(-k[0] * x[0], -k[1] * x[1] * x[0])
    
x = cs.MX.sym('x', 2)
t = cs.MX.sym('t')
k = cs.MX.sym('k', 2)
ode = {}  
ode['x'] = x  # states
ode['p'] = k
ode['t'] = t
ode['ode'] = simplerSim(x, t, k)

options = {}
options['grid'] = np.linspace(0,1)
t_iter = np.linspace(0, 1)

integrator = cs.integrator('F', 'cvodes', ode, options)

x0 = np.array([1, 2])
k_input = np.array([0.5, 0.2])

# Infinite difference; 
def getPertubedGrads(integrator, x0, k, epsilon_constant = None):
    n_k = k.shape[0]
    base_res = integrator(x0=copy.copy(x0), p=k)['xf'] #n_t, n_x
    grads = []
    for i_k in range(n_k):
        epsilon = epsilon_constant if epsilon_constant else k[i_k] * 0.01
        k_copy = copy.copy(k)
        k_copy[i_k] = k_copy[i_k] + epsilon
        pertubed_res = integrator(x0=copy.copy(x0), p=k_copy)['xf'] #n_t, n_x
        
        grad = (pertubed_res - base_res)/epsilon
        grads.append(grad)
    grads = np.array(grads) #n_k, n_t, n_x
    return grads

pertubed_grads = getPertubedGrads(integrator, x0, k_input) # n_k , n_x, n_t

# Casadi build in grad function
# I can not get grad for each x regard to each k
I_fwd = integrator.factory("I_fwd", ["x0", "p", "fwd:p"], ["fwd:xf"])
casadi_grads = np.array(I_fwd(x0=copy.copy(x0), p=k_input, fwd_p=np.array([1, 1]))['fwd_xf'])

n_show = 2
n_col = 2
n_row = (n_show//n_col) + (n_show % n_col != 0)

fig, axes = plt.subplots(n_row, n_col, figsize=(20, 8))
axes = axes.flatten()
for i in range(n_show):
    axes[i].plot(t_iter[1:], np.sum(pertubed_grads[i][:, :], axis = 0),label = "pertubed")
    axes[i].plot(t_iter[1:], casadi_grads[i],label = "casadi")
    axes[i].set_title("parameter {}".format(i))
    axes[i].legend()

``` 

Can someone explain to me what value I should use in fwd_p, please? Thank you so much in advance.

**Comment by jaeandersson** [comment]
`fwd_p` is the forward directional derivative seed corresponding to `p`. When you call `I_fwd = integrator.factory("I_fwd", ["x0", "p", "fwd:p"], ["fwd:xf"])`, you are creating a new function `I_fwd` defined by:

$$
(x_0, p, \texttt{fwd}_p) \rightarrow \frac{\partial x_f}{\partial p}(x0, p) * \texttt{fwd}_p
$$

The syntax for `factory` is the same for all CasADi functions, not just integrators. So you can create a function defined by symbolic expressions that has inputs `p` and `x0` and output `xf` and then use the same syntax.

**Reply by pvhau169** [reply]
Thank you for your response @jaeandersson,

I have a follow-up question regarding the use of I_fwd. Is there any way I can get the $\frac{\partial x_i}{\partial p_j}$? For now the I_fwd only returns $\frac{\partial x}{\partial p_j}$. I want $\frac{\partial x_i}{\partial p_j}$ to compare the results with the finite difference approximation.

**Reply by jaeandersson** [reply]
What do you mean by $\frac{\partial x_i}{\partial p_j}$? In general, if you want a derivative with respect to a specific variable, you should seed the corresponding variable. That is, let ``fwd_p`` be a unity vector with a 1 in position corresponding to $p_j$. You can also calculate the Jacobian, which gives all the entries.

**Reply by jaeandersson** [reply]
To get the Jacobian, you could do ``I_jac = integrator.factory("I_jac", ["x0", "p"], ["jac:xf:p"])``, if that helps

**Reply by pvhau169** [reply]
Thanks for your response @jaeandersson 

Let's me explain a bit what I mean by $\frac{\partial x}{\partial p_j}$.

With the finite difference approximation, for each pertubed parameter $p_j' = p_j + \epsilon_j$, I can get the finite difference by $\frac{integrator(x0, p') - integrator(x0, p)}{\epsilon_j} \in R^{n_x, n_t}$. Combining all the results from each perturbed parameter $p_j'$, I got pertubed_res $\in R^{n_x, n_t, n_k}$  where $n_x, n_t$, and $n_k$ are the number of states, timestamps and parameters.

With Casadi integrator.factory(x0, p , fwd_p), I understand that I can use fwd_p to activate specific parameters $p_j$. With fwd_p = 1, the result I got is casadi_res $\in R^{n_t, n_k}$. When I compare the casadi_res and pertubed_res, 
casadi_res = $\sum_{i} pertubedres_i$ where $i = [1, 2, 3, ... n_x]$.

How can I use Casadi to get the same results as the finite difference approximation approach?

**Reply by jaeandersson** [reply]
I think what you are looking for is the Jacobian of the state trajectories with respect to p. Try ``I_jac = integrator.factory("I_jac", ["x0", "p"], ["jac:xf:p"])``. The columns of the Jacobian will correspond to perturbations with respect to different parameters.

**Reply by jaeandersson** [reply]
Another option is to use the option "nfwd" when you construct the integrator, which was added in CasADi 3.6. Then the integrator will calculate an arbitrary number of forward sensitivities along with the evaluation. That's a bit of a different syntax though.

**Reply by pvhau169** [reply]
I tried I_jac = integrator.factory("I_jac", ["x0", "p"], ["jac:xf:p"]). It does give me the same results as I_fwd = integrator.factory("I_fwd", ["x0", "p", "fwd:p"], ["fwd:xf"])...

**Reply by jaeandersson** [reply]
Only if the dimension of p is 1 should they be the same

**Reply by pvhau169** [reply]
The dimension of p is $p \in R^{n_k}$ with n_k is the number of parameters. In my testing simulation n_k = 3. I used fwd_p = 1 for I_fwd and got the same answer.

**Reply by pvhau169** [reply]
To be more specific, my simulation has $n_s = 3, n_t = 5, n_k = 3$.

- I_fwd returns a 4x3 matrix (n_t-1, n_k)
{'fwd_xf': DM(
 [[-0.220624, -0.3894, -0.515467, -0.606531], 
  [-0.441248, -0.778801, -1.03093, -1.21306], 
  [-0.661873, -1.1682, -1.5464, -1.81959]])}
  
 - I_jac returns a 12x3 matrix but when I print it out, it's just like I_fwd results
 DM(sparse: 12-by-3, 12 nnz
 (0, 0) -> -0.220624
 (3, 0) -> -0.3894
 (6, 0) -> -0.515467
 (9, 0) -> -0.606531
 (1, 1) -> -0.441248
 (4, 1) -> -0.778801
 (7, 1) -> -1.03093
 (10, 1) -> -1.21306
 (2, 2) -> -0.661873
 (5, 2) -> -1.1682
 (8, 2) -> -1.5464
 (11, 2) -> -1.81959)

**Reply by jaeandersson** [reply]
fwd_xf is 3x4, i.e. (n_s, n_t -1) with your annotation, I suppose?

Each column in the Jacobian are the states for time point 0, all states for time point 1 etc. So your I_jac indicates that parameter 0 influences only state 0, parameter 1 influences only state 1 and parameter 2 influences only state 2. If your I_fwd calculation is the same, maybe you seeded all the parameter to 1? I.e. fwd_p = [1, 1, 1].

**Comment by pvhau169** [comment]
My bad, the fwd_xf is 3x4 which is (n_k, n_t-1).

Yes, I'm using fwd_p =1 which is the same as fwd_p = [1,1,1] as I wanted to activate all the parameters.

Compare the I_jac with I_fwd(fwd_p = 1),

I believe the second index corresponds to the parameter p, not the state. Index 0 influences the parameter 0, index 1 influences the parameter 1, etc...

Again I still don't know how to use the I_jac to achieve the same result as finite difference approach (n_s, n_t, n_k). When I check the shape of I_jac return (12x3) I assume it is (n_s * (n_t-1), n_k). But when I print it out it's just a sparse matrix where the results are same as I_fwd(fwd_p = 1).

**Reply by jaeandersson** [reply]
Right, the dimension is ns *(n_t-1) x n_k. When you take a Jacobian in CasADi, the elements get flattened, since CasADi doesn't have higher order tensors. So the column of the Jacobian corresponds to the flattened "xf" (which is ns x(n_t-1)). It's not correct that the results are the same as "I_fwd(fwd_p = 1)". What is true is that the nonzero entries are the same, which is due to particular structure of your problem where x[0] and x[1] are independent from each other and depend on different parameters.

**Reply by jaeandersson** [reply]
If you wanted to get the 3-rd order tensor, you can convert the CasADi sparse matrix to a numpy third-order tensor. numpy.reshape(I_jac.full(), [ns, nt-1, nk], order = 'F').

Also note that you are using deprecated syntax. If you are using CasADi 3.6, you should get a warning asking you to provide the time grid when calling cs.integrator instead of providing a "grid" option. If you are using an older version of CasADi than 3.6, you won't get a deprecation warning, but the result will probably not be correct if it at all runs (support for sensitivity analysis with multiple time points was added in CasADi 3.6).

**Reply by pvhau169** [reply]
Got it. That makes perfect sense now. Thank you so much for your support @jaeandersson

**Reply by jaeandersson** [reply]
You're welcome! On closer inspection, I see that x[0] and x[1] are not quite independent. But the rest of the comments still hold.

**Comment by pvhau169** [comment]
I got your point. I defined my test simulation as "cs.vertcat(-k[0] * x[0], -k[1] * x[1], -k[2] * x[2])" so I believe x[0] and x[1] should independent. Casadi returned me the independent grad but the finite difference approach didn't.
